tom_nonlocalizedevents/healpix_utils.py

```diff
@@ -158,28 +159,38 @@ def create_localization_for_skymap(nonlocalizedevent: NonLocalizedEvent, skymap_
         area_50, area_90 = get_confidence_regions(skymap)
 
         with transaction.atomic():
-            localization = EventLocalization.objects.create(
-                nonlocalizedevent=nonlocalizedevent,
-                distance_mean=distance_mean,
-                distance_std=distance_std,
-                skymap_version=skymap_version,
-                skymap_hash=skymap_uuid,
-                skymap_url=skymap_url,
-                area_50=area_50,
-                area_90=area_90,
-                date=date
-            )
-            for i, row in enumerate(skymap):
-                # This is necessary to make sure we don't get an underflow error in postgres
-                # when operating with the probdensity float field
-                probdensity = row['PROBDENSITY'] if row['PROBDENSITY'] > sys.float_info.min else 0
-                SkymapTile.objects.create(
-                    localization=localization,
-                    tile=uniq_to_bigintrange(row['UNIQ']),
-                    probdensity=probdensity,
-                    distance_mean=row_dist_mean[i] if row_dist_mean is not None else 0,
-                    distance_std=row_dist_std[i] if row_dist_std is not None else 0
+            try:
+                localization, is_new = EventLocalization.objects.get_or_create(
+                    nonlocalizedevent=nonlocalizedevent,
+                    skymap_hash=skymap_uuid,
+                    defaults={
+                        'distance_mean': distance_mean,
+                        'distance_std': distance_std,
+                        'skymap_version': skymap_version,
+                        'skymap_url': skymap_url,
+                        'area_50': area_50,
+                        'area_90': area_90,
+                        'date': date
+                    }
                 )
+                if not is_new:
+                    # This is added to protect against race conditions where the localization has already been added
+                    return localization
+                for i, row in enumerate(skymap):
+                    # This is necessary to make sure we don't get an underflow error in postgres
+                    # when operating with the probdensity float field
+                    probdensity = row['PROBDENSITY'] if row['PROBDENSITY'] > sys.float_info.min else 0
+                    SkymapTile.objects.create(
+                        localization=localization,
+                        tile=uniq_to_bigintrange(row['UNIQ']),
+                        probdensity=probdensity,
+                        distance_mean=row_dist_mean[i] if row_dist_mean is not None else 0,
+                        distance_std=row_dist_std[i] if row_dist_std is not None else 0
+                    )
+            except IntegrityError as e:
```

Comment:
But just in case, this is a fallback to catch a uniqueness constraint issue and return the localization that must exist then